dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,837 @@
+        }
+        SFGAOF att = MapFSToSFAttributes(item->GetFSAttributes()) | SFGAO_ISSLOW;
+        remain &= att & ~(FLATFOLDER ? SFGAO_FOLDER : 0);
+    }
+    *rgfInOut = remain;
+    return hr;
+}
+
+IFACEMETHODIMP CCabFolder::GetUIObjectOf(HWND hwndOwner, UINT cidl, PCUITEMID_CHILD_ARRAY apidl, REFIID riid, UINT *prgfInOut, LPVOID *ppvOut)
+{
+    HRESULT hr = E_NOINTERFACE;
+    if (riid == IID_IExtractIconA || riid == IID_IExtractIconW)
+    {
+        if (cidl != 1)
+            return E_INVALIDARG;
+        CABITEM *item = CABITEM::Validate(apidl[0]);
+        if (!item)
+            return E_INVALIDARG;
+
+        DWORD attr = item->GetFSAttributes();
+        return SHCreateFileExtractIconW(item->GetName(), attr, riid, ppvOut);
+    }
+    else if (riid == IID_IContextMenu && cidl)
+    {
+        LPFNDFMCALLBACK func = ItemMenuCallback;
+        IContextMenu **ppCM = (IContextMenu**)ppvOut;
+        return CDefFolderMenu_Create2(NULL, hwndOwner, cidl, apidl, this, func, 0, NULL, ppCM);
+    }
+    else if (riid == IID_IDataObject && cidl)
+    {
+        // Note: This IDataObject is only compatible with IContextMenu, it cannot handle drag&drop of virtual items!
+        return CIDLData_CreateFromIDArray(m_CurDir, cidl, apidl, (IDataObject**)ppvOut);
+    }
+    return hr;
+}
+
+IFACEMETHODIMP CCabFolder::MessageSFVCB(UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    switch (uMsg)
+    {
+        case SFVM_WINDOWCREATED:
+            m_ShellViewWindow = (HWND)wParam;
+            return S_OK;
+        case SFVM_WINDOWCLOSING:
+            m_ShellViewWindow = NULL;
+            return S_OK;
+    }
+    return E_NOTIMPL;
+}
+
+IFACEMETHODIMP CCabFolder::GetIconOf(PCUITEMID_CHILD pidl, UINT flags, int *pIconIndex)
+{
+    if (CABITEM *item = CABITEM::Validate(pidl))
+    {
+        int index = MapPIDLToSystemImageListIndex(this, pidl, flags);
+        if (index == -1 && item->IsFolder())
+            index = (flags & GIL_OPENICON) ? SIID_FOLDEROPEN : SIID_FOLDER;
+        if (index != -1)
+        {
+            *pIconIndex = index;
+            return S_OK;
+        }
+    }
+    return S_FALSE;
+}
+
+static HRESULT GetFsPathFromIDList(PCIDLIST_ABSOLUTE pidl, PWSTR pszPath)
+{
+    BOOL ret = SHGetPathFromIDListW(pidl, pszPath);
+    if (!ret && ILIsEmpty(pidl))
+        ret = SHGetSpecialFolderPathW(NULL, pszPath, CSIDL_DESKTOPDIRECTORY, TRUE);
+    return ret ? S_OK : HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);
+}
+
+static int CALLBACK FolderBrowseCallback(HWND hwnd, UINT uMsg, LPARAM lParam, LPARAM lpData)
+{
+    WCHAR buf[MAX_PATH];
+    switch (uMsg)
+    {
+        case BFFM_INITIALIZED:
+        {
+            if (LoadStringW(g_hInst, IDS_EXTRACT, buf, _countof(buf)))
+            {
+                // Remove leading and trailing dots
+                WCHAR *s = buf, *e = s + lstrlenW(s);
+                while (*s == '.') ++s;
+                while (e > buf && e[-1] == '.') *--e = UNICODE_NULL;
```

Comment:
This should probably be `e > s`.

Reply:
Probably but it would only matter if the text was nothing but periods. Still a good catch. In this specific case, the dots are either on the left or on the right but not both.